File: apps/macos/Sources/CornerTasks/Crypto/AccountManager.swift
import Foundation

protocol MnemonicProvider {
    func save(_ mnemonic: String) throws
    func load() throws -> String?
    func clear() throws
}

struct KeychainMnemonicProvider: MnemonicProvider {
    func save(_ mnemonic: String) throws { try MnemonicStore.save(mnemonic) }
    func load() throws -> String? { try MnemonicStore.load() }
    func clear() throws { try MnemonicStore.clear() }
}

enum AccountError: Error, Equatable {
    case invalidMnemonic
}

/// Holds the user's BIP-39 mnemonic and derived DID. Exposes `did` for UI binding;
/// the mnemonic is kept in memory only after a generate / import / load and is the
/// source for "Show mnemonic" / "Show QR code" reveals.
///
/// The Keychain is **not** queried on construction — opening the macOS app while
/// cloud sync is off must not produce a Keychain authorisation prompt. Callers must
/// invoke `loadIfPresent()` explicitly when the user has signalled intent to see
/// account info (opening Settings, expanding "Show mnemonic" / "Show QR code", or
/// the sync engine starting up). `loadIfPresent` is idempotent.
final class AccountManager: ObservableObject {
    @Published private(set) var did: String?
    private(set) var identity: Identity?
    private(set) var mnemonic: String?

    private let store: MnemonicProvider

    init(store: MnemonicProvider = KeychainMnemonicProvider()) {
        self.store = store
    }

    var hasKey: Bool { identity != nil }

    /// Reads the mnemonic from the configured store and populates `did`/`identity`/
    /// `mnemonic` if one is present. No-op once a mnemonic has already been loaded
    /// in this process — the Keychain is queried at most once per session per manager.
    func loadIfPresent() {
        if mnemonic != nil { return }
        guard let m = try? store.load(), Mnemonic.validate(m) else { return }
        applyMnemonic(m)
    }

    @discardableResult
    func generateNew() throws -> String {
        let m = try Mnemonic.generate()
        try store.save(m)
        applyMnemonic(m)
        return m
    }

    @discardableResult
    func importMnemonic(_ raw: String) throws -> String {
        let normalised = Self.normalise(raw)
        guard Mnemonic.validate(normalised) else { throw AccountError.invalidMnemonic }
        try store.save(normalised)
        applyMnemonic(normalised)
        return normalised
    }

    func forget() throws {
        try store.clear()
        mnemonic = nil
        identity = nil
        did = nil
    }

    static func normalise(_ raw: String) -> String {
        raw.lowercased()
            .components(separatedBy: .whitespacesAndNewlines)
            .filter { !$0.isEmpty }
            .joined(separator: " ")
    }

    private func applyMnemonic(_ m: String) {
        let seed = Mnemonic.toSeed(m)
        guard let id = try? Identity.fromBip39Seed(seed) else { return }
        mnemonic = m
        identity = id
        did = id.accountDid
    }
}

File: apps/web/src/crypto/AccountManager.ts
// SENSITIVE: holds the user's BIP-39 mnemonic in memory after a generate / import / load.
// Mirrors apps/macos/Sources/CornerTasks/Crypto/AccountManager.swift.

import * as mnemonic from './mnemonic';
import { fromBip39Seed, type Identity } from './identity';
import { MnemonicStore } from './MnemonicStore';

export class InvalidMnemonicError extends Error {
  constructor() {
    super('Invalid BIP-39 mnemonic');
    this.name = 'InvalidMnemonicError';
  }
}

export const normaliseMnemonic = (raw: string): string =>
  raw
    .toLowerCase()
    .split(/\s+/)
    .filter((w) => w.length > 0)
    .join(' ');

export interface AccountState {
  did: string | null;
  mnemonic: string | null;
  identity: Identity | null;
}

export class AccountManager {
  private state: AccountState = { did: null, mnemonic: null, identity: null };
  private listeners = new Set<(s: AccountState) => void>();

  private constructor(private store: MnemonicStore) {}

  static async open(store?: MnemonicStore): Promise<AccountManager> {
    const s = store ?? (await MnemonicStore.open());
    const m = new AccountManager(s);
    await m.loadIfPresent();
    return m;
  }

  get did(): string | null { return this.state.did; }
  get mnemonic(): string | null { return this.state.mnemonic; }
  get identity(): Identity | null { return this.state.identity; }
  get hasKey(): boolean { return this.state.identity !== null; }

  subscribe(fn: (s: AccountState) => void): () => void {
    this.listeners.add(fn);
    fn(this.state);
    return () => { this.listeners.delete(fn); };
  }

  async loadIfPresent(): Promise<void> {
    const m = await this.store.load();
    if (!m || !mnemonic.validate(m)) return;
    await this.applyMnemonic(m);
  }

  async generateNew(): Promise<string> {
    const m = mnemonic.generate();
    await this.store.save(m);
    await this.applyMnemonic(m);
    return m;
  }

  async importMnemonic(raw: string): Promise<string> {
    const normalised = normaliseMnemonic(raw);
    if (!mnemonic.validate(normalised)) throw new InvalidMnemonicError();
    await this.store.save(normalised);
    await this.applyMnemonic(normalised);
    return normalised;
  }

  async forget(): Promise<void> {
    await this.store.clear();
    this.set({ did: null, mnemonic: null, identity: null });
  }

  private async applyMnemonic(m: string): Promise<void> {
    const seed = await mnemonic.toSeed(m);
    const id = await fromBip39Seed(seed);
    this.set({ did: id.accountDid, mnemonic: m, identity: id });
  }

  private set(s: AccountState): void {
    this.state = s;
    for (const fn of this.listeners) fn(s);
  }
}
